### ISSUE TITLE
Request: extending `unnecessary-assign` (`RET504`) to account for `+=` (`__iadd__`)

### ISSUE THREAD

**Issue by jamesbraza** [astral-sh/ruff#9327]
With `ruff==0.1.9`'s [`unnecessary-assign`](https://docs.astral.sh/ruff/rules/unnecessary-assign/) (`RET504`), this goes undetected as having extra steps:

```python
def foo(arg: int) -> int:
    arg += 1
    return arg
```

It would be cool to extend this rule to request:

```python
def foo(arg: int) -> int:
    return arg + 1
```

There is a side effect here, it moves from invoking `__iadd__` to `__add__`. So perhaps an autofix for this should be considered an "unsafe" fix

**Comment by evanrittenhouse** [comment]
Feel free to assign this to me

**Comment by Skylion007** [comment]
I am linking a comment I made on the PR about how this may lead to undesirable behavior / unsafe edits: https://github.com/astral-sh/ruff/pull/9348#issuecomment-1873410403

**Comment by jamesbraza** [comment]
Yeah thanks for pointing that out. `__iadd__` can return an input mutable arg, while `__add__` will return a copy. I had alluded to this in the original post, thanks for making it more explicit.

I think a good approach is:

- Immutable return: autofix is "safe"
- Otherwise: autofix is "unsafe"

It looks like that's what @evanrittenhouse sort of proposed in his PR (and thanks for making the PR 👍)

**Comment by Skylion007** [comment]
Yeah, there is a question of fix safety, but there is also a question of false positives here since they may entirely different ops that could have different behaviors more subtly (even without side effects).

**Comment by jamesbraza** [comment]
> entirely different ops that could have different behaviors more subtly (even without side effects)

@Skylion007 I am not following what you meant by this, can you give an example?

Also, you do bring up the idea of other ops. It would be good to have this RET504 extension apply to other ops like `*=`, `-=`, `/=`, etc. May be worth adding to the test cases for that PR

**Comment by Skylion007** [comment]
Well, for instance, it's rare but a class have `__iadd__` implemented but not `__add__. And I was referring to the other inplace operator variants.

**Comment by evanrittenhouse** [comment]
> other ops like *=, -=, /=, etc

FWIW this operates on the `StmtAugAssign` node, so those operators will be handled automatically
